[sixthou] - Step1 PR 리뷰 부탁드립니다!

sixthou opened the pull request:
안녕하세요~🙌

1단계 첫번째 PR 드립니다~!
JPA를 새롭게 배우며 진행하니 어렵지만 열심히 한번 배워보겠습니다💪

이번 미션 진행하면서 궁금한점 질문드립니다!

1. 테스트 범위를 어떻게 설정해야 할까요?
 - 이전 미션을 진행하면서는 기존 코드가 없이 처음부터 작성을 했는데, 이번 미션은 기본적으로 주어지는 코드들이 있어 어디를 어떻게 테스트 해야하는지 판단하는게 어려웠습니다. 미션으로 주어진 엔티티 매핑과 관련된 부분만 테스트를 작성햇는데요. Answer, User와 같은 엔티티들의 메서드 중 일부와 Repository의 기본 CRUD, 제공된 조회 메서드들에 대해 테스트를 작성했습니다. 진행하다보니 모든 Repository에 대해서 CRUD 테스트를 작성하는게 벅차게 느껴지기도 하는데 테스트 범위를 어떻게 설정해야 할까요?

2.  assertall 사용 관련
- assertall을 이번 미션에서 새로 알게되어 적용을 해보았습니다. 테스트 작성시 assertall을 통해서 하나의 테스트로 테스트틀 진행하는것과 테스트를 분리해서 작성해야 하는지 사이에서 고민이 되더라구요. assertall을 Repository 저장 결과 확인을 위한 테스트 작성에만 활용하는게 아니라 QuestionRepositoryTest의 find_by_id_and_deleted_false 테스트 처럼 사용하는 것도 괜찮은 테스트 방법인지가 궁금합니다!

3. Repository 값 변경 관련된 테스트
- Repository에 테스트를 작성후 로그상 실행되는 쿼리를 봤는데 값 변경이나 삭제와 같이 영속성 컨테이너안에서 쓰기지연이 되는경우 insert -> delete가 아닌 insert만 진행되던데 제가 의도한 방향과 쿼리가 다르게 실행된다고 해도 과정을 통해 결과값 확인을 위해 테스트를 작성하는게 옳게 테스트를 작성한 것인지 궁금합니다.

첫번째 PR부터 질문을 너무 많이 드린것 같네요ㅜㅜ
이번 미션동안 잘 부탁드립니다!!

sixthou committed:
docs : README 작성

sixthou committed:
feat : jpa 옵션 추가 설정

sixthou committed:
feat : BaseEntity 생성
- 생성일자, 업데이트일자 공통 Entity생성 및 자동 저장을 위해 Auditing 설정
- Auditing 기능 화성화

sixthou committed:
feat : Answer 엔티티 맵핑
- BaseEntity 상속 처리

sixthou committed:
feat : DeleteHistory 엔티티 매핑

sixthou committed:
feat : Question 엔티티 매핑
- BaseEntity 상속처리

sixthou committed:
feat : User 엔티니매핑
- BaseEntity 상속처리

sixthou committed:
style : 컨벤션 정리

sixthou committed:
docs : Answer, AnswerRepository 테스트 목록 작성

sixthou committed:
test : Answer 테스트 작성

sixthou committed:
test : AnswerRepositoryTest 테스트 작성

sixthou committed:
test : DeleteHistory 테스트 작성

sixthou committed:
test : DeleteHistoryRepositoryTest 테스트 작성
- DeleteHistoryTest - 테스트용 전역 변수 선언
- DeleteHistory - 조회를 위한 getId 메서드 추가

sixthou committed:
docs : Question, QuestionRepository 테스트 항목 작성

sixthou committed:
test : Question 테스트 작성

sixthou committed:
test : QuestionRepository 테스트 작성

sixthou committed:
docs : User 관련 테스트 작성

sixthou committed:
test : UserTest 테스트 작성

sixthou committed:
feat : properties 로깅 옵션 추가
- sql 실행시 실 변수 확인 가능하도록 추가

sixthou committed:
test : UserRepositoryTest 테스트 작성

ohtaeg commented:
> 미션으로 주어진 엔티티 매핑과 관련된 부분만 테스트를 작성햇는데요. Answer, User와 같은 엔티티들의 메서드 중 일부와 Repository의 기본 CRUD, 제공된 조회 메서드들에 대해 테스트를 작성했습니다. 

미션에 맞게 테스트를 잘 작성해주셨어요 👍 👍 

> 진행하다보니 모든 Repository에 대해서 CRUD 테스트를 작성하는게 벅차게 느껴지기도 하는데 테스트 범위를 어떻게 설정해야 할까요?

사실 이번 미션의 목적은 JPA 학습과 활용이 목적이기 때문에 학습을 위해 작성하시는 Repository의 테스트 양이 많을 수 밖에 없으셨을거에요!
벅차게 느끼신게 당연하며 그러기 위해서 코치분들께서 미션을 준비해주셨어요 👍 
미션마다 요구사항과 환경이 다르기에 테스트 범위에 대해 말씀 드리기가 애매하지만 
해당 미션에서의 테스트 범위는 말 그대로 엔티티와 `@DataJpaTest`를 활용한 Repository가 범위인데 잘해주셨습니다!
다른 미션에서 테스트할 때는 또 다른 느낌을 받으실 수 있으실거에요 :)
꼼꼼하게 테스트 작성하시느라 고생 많으셨습니다 💯

ohtaeg commented:
> assertall을 이번 미션에서 새로 알게되어 적용을 해보았습니다. 테스트 작성시 assertall을 통해서 하나의 테스트로 테스트틀 진행하는것과 테스트를 분리해서 작성해야 하는지 사이에서 고민이 되더라구요. assertall을 Repository 저장 결과 확인을 위한 테스트 작성에만 활용하는게 아니라 QuestionRepositoryTest의 find_by_id_and_deleted_false 테스트 처럼 사용하는 것도 괜찮은 테스트 방법인지가 궁금합니다!

find_by_id_and_deleted_false에서도 잘해주셨고 save() 테스트에서도 잘 활용해주셨습니다 👍 
하나의 시나리오(기능)을 검증하는데 여러 검증문이 필요할 경우 사용해주시면 됩니다 :)
테스트를 분리한다는 것은 서로 다른 시나리오를 테스트 하는 것으로써, 
다른 시나리오가 아닌 상태에서 여러 검증이 필요하면 해주신 것처럼 활용해주시면 될 것 같아요 :)

게다가 각각의 검증문과 assertAll로 묶인 검증문들의 차이는 예외 발생했을 때의 장점이 있는데요!
이 부분에 대해 자세하게 설명해주신 블로그가 있어 링크로 대체 드리오니 한번 읽어보시면 좋을 것 같아요! 🙇 
https://escapefromcoding.tistory.com/358

ohtaeg commented:
> Repository에 테스트를 작성후 로그상 실행되는 쿼리를 봤는데 값 변경이나 삭제와 같이 영속성 컨테이너안에서 쓰기지연이 되는경우 insert -> delete가 아닌 insert만 진행되던데 제가 의도한 방향과 쿼리가 다르게 실행된다고 해도 과정을 통해 결과값 확인을 위해 테스트를 작성하는게 옳게 테스트를 작성한 것인지 궁금합니다.

우선 전체적으로 테스트를 잘 작성해주셨습니다 :)
옳게 작성해주셨어요 👍 
여기서 말씀해주신대로 insert() 및 delete() 관련해서 쓰기지연 저장소에 저장되어 있을텐데 왜 delete()는 실행되지 않았을까요 :) ?
이건 학습을 위해 제가 답변을 드리는 것보다 한번 고민해보시는 편이 공부가 될 것 같아요!

아래 테스트 코드를 기준으로 생각해보시면 좋을 것 같아요!

```
    @DisplayName("저장된 Answer을 삭제할 수 있다.")
    @Test
    void delete() {
        Answer actual = answerRepository.save(AnswerTest.A1);

        answerRepository.delete(actual);
        Optional<Answer> findAnswer = answerRepository.findById(actual.getId());

        assertThat(findAnswer).isNotPresent();
    }
```

delete() 와 findById 사이에 flush()를 수행하면 delete 쿼리가 찍히는지 확인해보시면 좋을 것 같아요!
확인해보신 후 그래도 감이 안오신다면 아래 힌트를 참고하시면 좋을 것 같아요 :)

<details>
<summary>힌트</summary>
findById가 아닌 findByContents()로 한번 조회를 해보시도록 하여
delete() 쿼리가 발생하는지 안하는지 확인해보시면 좋을 것 같습니다 :) <br>
ID 기반으로 조회하게 되면 내부적으로 어떤 일이 발생하는지 공부해주신 후 설명 해주세요!!
</details>

ohtaeg commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,17 +1,32 @@
+    @Column(nullable = false)
     private boolean deleted = false;
 
+    protected Answer() {
```
protected 👍

ohtaeg commented on src/main/java/qna/Application.java:
```diff
@@ -2,7 +2,9 @@
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.data.jpa.repository.config.EnableJpaAuditing;
 
+@EnableJpaAuditing
```
JPA Auditing을 활성화 👍

ohtaeg commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -2,25 +2,46 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
+    @Column(columnDefinition = "varchar(255)")
```
잘해주셨어요 👍 
`@Enumerated(EnumType.STRING)`으로도 할 수 있습니다 :)

sixthou replied:
dml에 맞춰 진행하다 보니 enum 타입인걸 놓쳤네요😭

ohtaeg commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -2,25 +2,46 @@
+            return false;
+        }
         DeleteHistory that = (DeleteHistory) o;
         return Objects.equals(id, that.id) &&
```
엔티티의 동등성을 비교하고자 할 때 모든 값을 비교 하도록 해주신 의도가 있으신가요 :) ?
크게 상관은 없지만 엔티티 같은 경우는 id가 유니크함을 보장하기에 id로만 판단해도 괜찮지 않을까 하는데
천호님은 어떻게 생각하시나요 😄

sixthou replied:
두개의 객체를 비교할때 모든 값을 비교해줘야 한다고 생각했는데. jpa에서 id가 유니크함을 보장해준다면 굳이 다른 값까지 비교할 필요는 없을것 같아요

ohtaeg commented on src/test/java/qna/domain/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,94 @@
+package qna.domain;
```
개인적으로 Entity와 Repository 테스트는 패키지 분리해서 작성하는 것이 좋다고 생각하는데요!
domain 패키지에 같이 해주신 이유가 있으신가요 :) ?

sixthou replied:
Entity와 Repository의 패키지를 분리하니 더 명확하게 관리가 가능할 것 같네요!

ohtaeg commented on src/test/java/qna/domain/DeleteHistoryRepositoryTest.java:
```diff
@@ -0,0 +1,39 @@
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+
+@DataJpaTest
+class DeleteHistoryRepositoryTest {
```
전체적으로 테스트를 꼼꼼하게 잘 작성해주셨어요 💯

ohtaeg commented on src/main/java/qna/domain/User.java:
```diff
@@ -1,19 +1,30 @@
+import qna.UnAuthorizedException;
 
-public class User {
+@Entity
```
전체적으로 엔티티 요구사항을 잘 만족시켜주셨습니다 👍

ohtaeg reviewed (changes requested):
안녕하세요 천호님!
두 번째 JPA 미션을 함께할 리뷰어 오태경이라고 합니다!!
함께하게 되어 영광이며, 잘 부탁드리겠습니다 🙇

엔티티 매핑을 꼼꼼히 해주려고 하신 점과 테스트 코드를 꼼꼼하게 잘 작성해주셨어요 💯
주셨던 질문에 대해 답변 및 소소한 코멘트를 드렸는데 확인해주시면 감사하겠습니다 :)

다음 미션을 진행하기 전에 주셨던 질문에 대해 학습하시고 진행한다면 
다음 미션을 수월하게 할 수 있을 것 같아 Request changes를 하였습니다 :)
답변을 확인해주시고 설명을 해주시면 감사하겠습니다 😄

sixthou committed:
refactor : DeleteHistory 피드백 반영
- ContentType : Enum타입에 맞게 변경
- equals() : 엔티티의 경우 id 값의 유니크함이 보장되기 때문에 id값으로만 변경되게 수정

sixthou committed:
refactor : Repository 패키지 분리

sixthou commented:
친절하고 자세하게 답변주셔서 매우 감사드립니다~🙇🙇

3번 질문에 대해 코멘트 남겨주신부분 작성해 보아요!

- 영속성 컨텍스트는 엔티티를 식별자(ID) 값으로 구분한다.
- 플러시를 하는 순간 데이터베이스에 반영한다.
- 조회시 우선 1차 캐시에서 식별자(ID) 값으로 엔티티를 찾고, 찾는 엔티티가 있는 경우에는 DB를 조회하지 않는다. 없을 경우에는 DB를 조회해 엔티티를 생성하고, 1차 캐시에 저장후 엔티티를 반환한다.
- 커밋직전까지 DB에 엔티티를 저장하지 않고, 내부 쿼리 저장소에 모아둔고 커밋시 모아둔 쿼리를 보낸다.

위를 바탕으로 3가지 케이스별로 내부 동작을 유추해봤습니다.

```java
@Test
void delete1() {
    // insert 실행
    Answer actual = answerRepository.save(AnswerTest.A1);
    // 1. ID값을 얻기 위해 SQL 실행.
    answerRepository.delete(actual);
    // 2. 1차캐시에 ID는 남아있으나 엔티니는 제거됨.
    Optional<Answer> findAnswer = answerRepository.findById(actual.getId());
    // 3. 1차캐시에서 ID로 조회.
    assertThat(findAnswer).isNotPresent();
}
```

```java
@Test
void delete2() {
    // insert -> delete -> select
    Answer actual = answerRepository.save(AnswerTest.A1);
    // 1. ID값을 얻기 위해 SQL 실행.(insert)
    answerRepository.delete(actual);
    // 2. 1차캐시에서 ID는 남아있으나 엔티니는 제거됨.
    answerRepository.flush();
    // 3.  쓰기 지연 SQL 저장소 저장소 쿼리 실행. -> 1차캐시 id 삭제 (delete)
    Optional<Answer> findAnswer = answerRepository.findById(actual.getId());
    // 4. 1차캐시에 조회 -> id 값이 없으니깐 db에 select 실행. (select)
    assertThat(findAnswer).isNotPresent();
}
```
```java
@Test
void delete3() {
    // insert -> delete -> select
    Answer actual = answerRepository.save(AnswerTest.A1);
    // 1. ID값을 얻기 위해 SQL 실행. (insert)
    answerRepository.delete(actual);
    // 2. 1차캐시에 ID는 남아있으나 엔티니는 제거됨.
    Optional<Answer> byContents = answerRepository.findByContents(AnswerTest.A1.getContents());
    // 3. 1차 캐시에서 contents로 조회
    //      -> 없으니깐 db 조회를 해야함
    //      -> 쓰기 지연 SQL 저장소 저장소 쿼리 실행하여 변경 반영 (delete)
    //      -> 조회(select)
    assertThat(byContents).isNotPresent();
}
```

ohtaeg commented on src/main/java/qna/repository/AnswerRepository.java:
```diff
@@ -1,9 +1,10 @@
-package qna.domain;
+package qna.repository;
```
피드백 반영 👍

ohtaeg commented:
너무 잘해주셨어요 💯 👍 👍 
잘 정리해주셔서 추가로 말씀드릴 부분이 없네요!!
다음 미션도 기대하겠습니다 😄

ohtaeg reviewed (approved):
피드백 반영 및 설명을 잘해주셨어요 💯 
다음 미션도 화이팅입니다 🔥 
갑자기 추워진 날씨로 인해 감기 걸리지 않도록 조심하세요 :)
